### ISSUE TITLE
Wrong handling of TImestamp-diffs

### ISSUE THREAD

**Issue by mschirmacher** [jOOQ/jOOQ#6335]
We are using jooq 3.9.0 and Postgresql 9.5.
Now we have a table (lets call it `table`) having a column (lets call it `column`) of type `timestamp without time zone`.
I intended to write a query like 
`DSL.currentTimestamp.subtract(TABLE.COLUMN).lessThan( … ).`
Now jooq expects the argument of `lessThan` to be of type java.sql.Timestamp. Yet PostgreSQL expects an interval type.
So assuming i want all `column`s being older then one day writing `…lessThan(new Timestamp(86400))` results in a BadGrammarException.
Correct would be `lessThan(new DayToSecond(1))` in this case

**Comment by lukaseder** [comment]
Thanks for pointing this out. Indeed, date subtraction is different from "ordinary" arithmetic subtraction as the result type of the subtraction differs from the operand types. In that sense, it is a bit unfortunate that the `+` and `-` operators were overloaded, specifically because `+` is usually associative (and mostly also commutative).

In any case, we got this wrong in our API, but there's no way we can keep type safety and fix this in the same move:

- If we want to keep type safety, we could no longer overload the method name, because Java doesn't support union types or reified generics
- If we drop type safety, we could let the result of subtract be of type `Field<Object>`, but that would be quite a loss of functionality (and backwards-incompatible).

We'll have to think about how to fix this. In the meantime, you could use `DSL.timestampDiff()` instead, which gets types correctly.

**Comment by mschirmacher** [comment]
Yeah indeed, `DSL.timestampDiff()` makes a good job here, too :+1:
